File: src/dotnet/Ansight.Tools.Reflection/ReflectionOptionsBuilderExtensions.cs
namespace Ansight.Tools.Reflection;

public static class ReflectionOptionsBuilderExtensions
{
    public static Options.OptionsBuilder WithReflectionTools(this Options.OptionsBuilder builder)
        => builder.WithReflectionTools(static _ => { });

    public static Options.OptionsBuilder WithReflectionTools(
        this Options.OptionsBuilder builder,
        Action<ReflectionToolsOptionsBuilder> configure)
    {
        ArgumentNullException.ThrowIfNull(builder);
        ArgumentNullException.ThrowIfNull(configure);

        var optionsBuilder = ReflectionToolsOptions.CreateBuilder();
        configure(optionsBuilder);
        return builder.WithReflectionTools(optionsBuilder.Build());
    }

    public static Options.OptionsBuilder WithReflectionTools(
        this Options.OptionsBuilder builder,
        ReflectionToolsOptions options)
    {
        ArgumentNullException.ThrowIfNull(builder);
        ArgumentNullException.ThrowIfNull(options);

        return builder.AddTools(new ITool[]
        {
            new ListReflectionRootsTool(options),
            new InspectObjectTool(options),
            new DescribeTypeTool(options),
            new SetMemberValueTool(options),
            new InvokeMethodTool(options)
        });
    }
}

File: src/android/ansight-tools-reflection/src/main/kotlin/ai/ansight/tools/reflection/ReflectionOptionsBuilderExtensions.kt
package ai.ansight.tools.reflection

import ai.ansight.runtime.AnsightOptionsBuilder

fun AnsightOptionsBuilder.withReflectionTools(): AnsightOptionsBuilder {
    return addTools(AndroidReflectionTools.create())
}

fun AnsightOptionsBuilder.withReflectionTools(
    options: AndroidReflectionToolsOptions,
): AnsightOptionsBuilder {
    return addTools(AndroidReflectionTools.create(options))
}

fun AnsightOptionsBuilder.withReflectionTools(
    configure: AndroidReflectionToolsOptionsBuilder.() -> Unit,
): AnsightOptionsBuilder {
    val builder = AndroidReflectionToolsOptions.createBuilder()
    builder.configure()
    return withReflectionTools(builder.build())
}
